doc/developer/design/20230512_mir_cost_model.md

```diff
@@ -0,0 +1,149 @@
+
+Our cost model should be in terms of the dimensions that are relevant to us:
+
+- Latency (rehydration time; latency per update)
+- Storage (memory footprint/implicit storage)
+- Scalability (to what degree can this query be spread across workers)
+
+Our cost model should be *executable* and *accurate:*
+
+- Our cost model will be code that runs on actual MIR, producing a symbolic expression that represents the cost of the query along various dimensions. A user can use `EXPLAIN WITH(cost)`  to predict costs.
+- Our cost model should be tested on a variety of loads to ensure that our estimates aren’t total nonsense.
+
+# Motivation
+[motivation]: #motivation
+
+## Goals
+
+- Reify what we learn from various benchmarks/experiments.
+    - https://github.com/MaterializeInc/materialize/issues/17883
+    - https://github.com/MaterializeInc/materialize/issues/16511
+    - https://github.com/MaterializeInc/materialize/issues/15244
+    - https://github.com/MaterializeInc/materialize/issues/13140
+    - https://github.com/MaterializeInc/materialize/issues/12069
+    - https://github.com/MaterializeInc/materialize/issues/9157
+- Offer visibility/predicability.
+    - https://github.com/MaterializeInc/materialize/issues/14663
+    - https://github.com/MaterializeInc/materialize/issues/13046
+- Open opportunities for principled/well-informed choices in the optimizer.
+
+## Non-goals
+
+- Identify all opportunities to use the cost model in the optimizer.
+- Refactor the optimizer to be focused on the cost model.
+- Implement dynamic re-optimization based on changes in input statistics.
+- Collect all relevant statistics in storage and make them available to the cost model.
+
+# Explanation
+[explanation]: #explanation
+
+## Proposal
+
+A cost model is an abstract interpretation of the query: rather than running the query on relations and getting new relation out, we run the query on a symbolic description of its input relations and get a symbolic description of the result out.
+
+The general approach can be the same as `mz_transform::Typecheck`: recursively walk the `MirRelationExpr` with some contextual information, returning a symbolic term characterizing the query.
+
+### Context information
+
+Traditional databases track cardinality data and statistics about various columns. We will be interested in other statistics, too. Some possible examples:
+
+- How frequently does data arrive?
+- Is the given input physically monotonic or append-only? (I.e., there are no retractions.)
+- At what rate do we expect new data to arrive?
+- Some kind of locality property?
+
+### Symbolic costs
+
+A symbolic cost is a polynomial over the features of the input relations. For example, a batch query that loops over relation `R` twice and then computes its cross product with relation `Q` will have cost `2|R| + |R|*|Q|`. A traditional database typically computes cost just in terms of overall batch query time, but our costs are multi-dimensional. In rough feasibility order:
```

Comment:
> A symbolic cost is a polynomial over the features of the input relations.

This is more of an implementation question, but when we implement this, will we (1) normalize the polynomial representing the cost of every sub-term or (2) produce a term that simplifies to a polynomial but reflects the shape of the input tree. If (1), is there a standard algorithm to have in mind here?

Reply:
We may not need to simplify at all, if we can fully substitute every variable. But in case we ever want to compare symbolic costs, I think we'll want to follow the algorithms in Joel S. Cohen’s _Computer Algebra and Symbolic Computation_.

There are algorithms for ordinary algebraic expressions and for "polynomial simplification with side relations", i.e., reducing a polynomial `Q(x1, ..., xn) = ...` with respect to equations `f1 = 0, ..., fm = 0`, where each equation `fj` is in terms of the variables `xi`. The former may suffice, but the latter will let us reduce a polynomial with partial information (i.e., `x1` and `x2` are equal, but we don't know their values; if we know their values, we can just substitute).

I don't have a clear sense of whether it's better to simplify as we go or just do it at the end, but I'd guess it's better to let things get big and shrink them later.